provider/inmemory.go

```diff
@@ -126,38 +131,50 @@ func (im *InMemoryProvider) validateChangeBatch(zone string, changes *plan.Chang
 	if !ok {
 		return ErrZoneNotFound
 	}
-	mesh := map[string]bool{}
+	mesh := map[string]map[string]bool{}
```

Comment:
This type looks too complicated to me. 
All the code below looks very complicated to read, because of that. I think you should create a type and some methods that works on top of it, to have it simpler